tests/testthat/test-return_linter.R

```diff
@@ -0,0 +1,841 @@
+    ),
+    return_linter()
+  )
+})
+
+test_that("Do not lint control statements (with return) on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+
+  expect_lint(
+    trim_some("
+      function() {
+        repeat {
+          cat(4)
+        }
+      }
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        if(x) {
+          return(4)
+        } else if(y) {
+          return(5)
+        } else {
+          return(6)
+        }
+      }
+    "),
+    NULL,
+    linter
+  )
+})
+
+test_that("Lint control statements (without return) on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+  msg <- rex::rex("All functions must have an explicit return().")
+
+  expect_lint(
+    trim_some("
+      function() {
+        while(x > 4) {
+          cat(4)
+          if(x < 4) {
+            return(x)
+          }
+        }
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        for(i in 1:10) {
+          cat(4)
+          if(i > 11) {
+            return(x)
+          }
+        }
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        if(x == 2L){
+          return(e)
+        }
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        if(x == 2L){
+          return(e)
+        } else if(x == 3L) {
+          cat(f)
+        }
+      }
+    "),
+    list(msg, line_number = 4L),
+    linter
+  )
+})
+
+test_that("Do not lint stop on end of function", {
+  expect_lint(
+    trim_some("
+      function() {
+        # Test
+        3 + 4
+        stop(1)
+      }
+    "),
+    NULL,
+    return_linter(use_implicit_returns = FALSE)
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        stop(1)
+      }
+    "),
+    NULL,
+    return_linter()
+  )
+})
+
+test_that("Do not lint stop on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+  msg <- rex::rex("All functions must have an explicit return().")
+
+  expect_lint(
+    trim_some("
+      function(x) {
+        switch(x, a = 1, 'b' = 2, '3' = 3, 4)
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function(x) {
+        switch(x, a = return(1), 'b' = stop(2), '3' = return(3), 4)
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        switch(
+          x,
+          a = return(1),
+          'b' = stop(2),
+          '3' = return(3)
+        )
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function(x) {
+        switch(x, a = return(1), 'b' = stop(2), '3' = return(3), stop('End'))
+      }
+    "),
+    NULL,
```

Comment:
This also looks off to me -- I would expect this to lint. For explicit returns, I think it's better to force "outer" return: `return(switch(...))`.